articles/logging.md

```diff
@@ -0,0 +1,75 @@
+To define the interface of the ROS2 logging system, it is helpful to consider all of the various tasks users will want to perform.
+From this list of tasks, we can describe the steps users will have to take in order to perform each one.
+Based on our experiences with ROS1, other logging systems, and community feedback, the logging system in ROS2 will support the following interactions:
+- creating loggers
+- setting log levels
+- outputting logs
+- defining output destinations
+- specifying a backend
+- implementing a custom backend
+- tailoring performance impact
+### Interaction Descriptions
+This section describes what each of the interactions in the list entails, and the steps users will have to take to complete them.
+#### Creating Loggers
+This interaction contitutes creating the objects that users will use to output information from their code.
+#### Setting Log Levels
+This is how users set the severity levels for the loggers in their code.
+There are a number of ways users might want to be able to change these levels, so this interaction is further broken down to describe each of those.
+##### Programatically
+##### Command Line
+##### Via ROS2 Service Call
+##### Environment Variable?
+#### Outputting Logs
+This is primary interaction where users output information from their code.
+#### Defining Output Destinations (Sinks)
+Users may want logged information printed to the screen, written to a file, published to a topic, or a combination of those. This is how users setup those preferences.
+#### Specifying A Backend
+There are a number of widely used third-party logging libraries already available which may better suit a users use-case or provide extra features they need.
+ROS2 logging provides a default implementation, as well as a couple alternatives.
+This defines how users choose the implementation that best suits their needs.
+#### Creating A Backend
+Users may have requirements that none of the included backends support, so ROS2 provides an interface for users to create custom backend implementations.
+#### Tailoring Performance Impact
+Adding a sophisticated logging system to a complex code base can have performance impacts and are often mostly useful during development and debugging.
+ROS2 provides a few ways to configure the logging system to eliminate this performance impact when the logs are not needed.
+This section describes how to make those configuration changes.
+
+
+
+## Existing infrastructure
+ROS2 
+### `rcutils`
+Contains macros for logging at several different levels and under various constraints
+### `rclcpp`
+Contains similar macros to those in `rcutils`, which call down to the macros in `rcutils`
+### `rcl_logging`
+Contains a logging interface and several implementations including log4cxx, spdlog, and a no-op implementation
+### `rcl`
+Contains infrastructure to tie the logger implementations defined in `rcl_logging` to the macros in `rcutils`
```

Comment:
You know, thinking about it further, I'm wondering if we need configuration for that to be in ROS 2 at all.  That is, if you are enabling debugging on an RMW vendor, then you are obviously trying to debug something in that particular vendor.  The vendors all have vendor-specific ways to enable debugging, so wouldn't it make sense to use one of those?

(this line of reasoning also questions the `rmw_set_log_severity` API at all)

Reply:
yeah, that's a valid question.

> That is, if you are enabling debugging on an RMW vendor, then you are obviously trying to debug something in that particular vendor.

This is true, but

> The vendors all have vendor-specific ways to enable debugging, so wouldn't it make sense to use one of those?

this also feels like an argument in favour of an interface like `rmw_set_log_severity`.

Reply:
good point. Some vendor-specific configuration is expected (like FastDDS's xml config files), so maybe logging can just be bundled with that?

The middleware WG is a good place to have a discussion about where the line should be drawn!

Reply:
FYI @clalancette @mlanting I added this discussion/question to today's middleware WG meeting agenda. Sorry for the short notice.

Reply:
We had a short discussion about this in the last middleware WG meeting (2021-05-19). I'd recommend checking out the meeting notes.

In short: `rmw_set_log_severity` has a purpose even if there are other ways to set an rmw/DDS implementation's logging level. We could use a "well-known"/special logger name to set the logging level through the CLI (using this syntax: https://docs.ros.org/en/rolling/Tutorials/Logging-and-logger-configuration.html#logger-level-configuration-command-line). However, we have to make sure that the rmw implementation does not produce log messages if they get filtered out at the ROS 2 level (in case it is set as a log consumer).